# **Maestría en Inteligencia Artificial Aplicada**

## **Curso: Inteligencia Artificial y Aprendizaje Automático**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

## **Actividad de las Semanas 5 y 6**
### **Problema de asignación de créditos: South German Dataset.**


## **Nombres y matrículas:**


🔹 Jossie Paola Jiménez Rivera   ➡️ A01795961

🔹 Javier Augusto Rebull Saucedo ➡️ A01795838

🔹 Nancy Teresa Zapién García    ➡️ A01795907



# **Parte I: Partición, análisis y pre-procesamiento de los datos.**

## **Ejercicio 1**

In [17]:
# Aquí deberás incluir todas las librerías que requieras durante esta actividad:

#Para manipular datos
import numpy as np
import pandas as pd

#Ejercicio 3.
from sklearn.model_selection import train_test_split

#Ejercicio 5.
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer

#Ejercicio 7.
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC
  #para conjuntos desbalanceados
from imblearn.over_sampling import SMOTE, KMeansSMOTE
from sklearn.model_selection import RepeatedStratifiedKFold, cross_validate
from sklearn.metrics import make_scorer
from imblearn.metrics import geometric_mean_score

In [18]:
# Si se desean comentar algunos de los Warnings.
import warnings
warnings.filterwarnings('ignore')

In [19]:
# Carga y renombra los nombres de las columnas del alemán al inglés y desplegamos
# de nuevo el DataFrame para ver el resultado obtenido:

# ************* Inlcuye aquí tu código:*****************************
# NOTA: Tanto aquí como en lo sucesivo, "None" puede indicar una o varias líneas de código.

from google.colab import drive
drive.mount('/content/drive')

import os
DIR='/content/drive/MyDrive/TecMTY/septiembre noviembre 2024/IA y ML/Semana 5 y 6'
os.chdir(DIR)

#lectura del .asc
df=pd.read_csv('SouthGermanCredit.asc', sep=' ')
#df

#Renombra columnas de Aleman a Ingles
columns=['status','duration','credit_history','purpose','amount','savings','employment_duration',
         'installment_rate','personal_status_sex','other_debtors','present_residence','property',
         'age','other_installment_plans','housing','number_credits','job','people_liable','telephone',
         'foreign_worker','credit_risk']
#guarda nuevos nombre de las columnas al df
df.columns=columns

# *********** Aquí termina la sección de agregar código *************

df.head().T

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


0     1    2     3     4
status                      1     1    2     1     1
duration                   18     9   12    12    12
credit_history              4     4    2     4     4
purpose                     2     0    9     0     0
amount                   1049  2799  841  2122  2171
savings                     1     1    2     1     1
employment_duration         2     3    4     3     3
installment_rate            4     2    2     3     4
personal_status_sex         2     3    2     3     3
other_debtors               1     1    1     1     1
present_residence           4     2    4     2     4
property                    2     1    1     1     2
age                        21    36   23    39    38
other_installment_plans     3     3    3     3     1
housing                     1     1    1     1     2
number_credits              1     2    1     2     2
job                         3     3    2     2     2
people_liable               2     1    2     1     2
telephone                   1     1    1     1     1
foreign_worker              2     2    2     1     1
credit_risk                 1     1    1     1     1

## **Ejercicio 2**

In [20]:
# Realiza a continuación una transformación para que la clase negativa (buen cliente)
# quede con el valor de 0 y la clase positiva (mal cliente) quede con el valor de 1.

# ************* Inlcuye aquí tu código:*****************************

df['credit_risk']=df['credit_risk'].replace({0:1,1:0})

# *********** Aquí termina la sección de agregar código *************

print(df['credit_risk'].value_counts())

credit_risk
0    700
1    300
Name: count, dtype: int64


## **Ejercicio 3**

In [21]:
# Realiza una partición solicitada de entrenamiento y prueba.
# Los nombres de los conjuntos deberán ser como se indican en los print de abajo:

# ************* Inlcuye aquí tu código:*****************************

X=df.drop(columns='credit_risk') #Características
y=df['credit_risk'] #Variable Salida

# El articulo sugiere un partición de (70:30), Entrenamiento 70% y Prueba 30%
Xtrain, Xtest, ytrain, ytest=train_test_split(X,y, test_size=0.3, stratify=y, random_state=1)

# *********** Aquí termina la sección de agregar código *************

# Mostremos las dimensiones de la partición generada:
print("Dimensiones:")
print("Entrenamiento:", Xtrain.shape, ytrain.shape)
print("Prueba:", Xtest.shape, ytest.shape)

# Y el porcentaje de cada clase de la variable de salida:
tmp = ytrain.sum()/ytrain.shape[0]
print("\nPorcentaje clases Positiva:%.2f%%, y Negativa:%.2f%%" % (100*(1-tmp),tmp*100))


Dimensiones:
Entrenamiento: (700, 20) (700,)
Prueba: (300, 20) (300,)

Porcentaje clases Positiva:70.00%, y Negativa:30.00%


### **Con base al porcentaje de los niveles de la variable de salida ¿podemos decir que tenemos un problema de datos desbalanceado? ¿Por qué?**



### ++++++++ Inicia la sección de agregar texto: +++++++++++

En base a los porcentajes obtenidos, podemos concluir que **si existe un desbalance con el conjunto de datos**. Ya que, del **100%** de todos nuestros datos se desgloza en un **70%** para la **Clase Positiva (mal cliente)** y un **30%** para la **Clase Negativa (buenos clientes)**. Este desbalance puede afecttar el rendimiento de los modelos de ML.

Al trabajar con datos desbalanceados, el modelo tiende a ser bueno en identificar la clase mayoritaria(0), mientras que no es tan preciso para predecir la clase minoritaria(1).

### ++++++++ Termina la sección de agregar texto: +++++++++++



## **Ejercicio 4**

In [22]:
# De acuerdo a la información de la Tabla 3 del artículo de la IEEE
# define las variables correspondientes que se indican a continuación:

# ************* Inlcuye aquí tu código:*****************************

# Variables numéricas:
lista_paper_num = ['duration', 'amount','age','peope_liable']
#4variables

# Variables ordinales:
lista_paper_ord = ['employment_duration','installment_rate','present_residence',
                   'property','number_credits','job']
#6 variables

# Variables nominales:
lista_paper_cat = ['status','credit_history','purpose','savings','personal_status_sex',
                   'other_debtors','other_installment_plans','housing','telephone','foreign_worker']#credit_risk, no se tomo encuenta
#10 variables

# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 5**

In [23]:
# Transformaciones que se aplicarán a las variables numéricas usando la clase Pipeline de sklearn:

# ************* Inlcuye aquí tu código:*****************************

# Variables numéricas:
numericas_pipe = Pipeline([("impute", SimpleImputer(strategy="median")),
                           ("standardize", MinMaxScaler())])
numericas_pipe_nombres = ['duration', 'amount','age', 'people_liable']

# Variables categóricas-Nominales:
nominales_pipe = Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                           ("encoder", OneHotEncoder(handle_unknown="ignore", drop = "first"))])
nominales_pipe_nombres = ['status','credit_history','purpose','savings','personal_status_sex',
                          'other_debtors','other_installment_plans','housing','telephone','foreign_worker']#quite credict_risk

# Variables categóricas-ordinales:
ordinales_pipe = Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                           ("encoder", OrdinalEncoder())])
ordinales_pipe_nombres = ['employment_duration','installment_rate','present_residence','property','number_credits','job']


# Conjuntas las transformaciones de todo tipo de variable y
# deja sin procesar aquellas que hayas decidido no transformar:

columnasTransformer = ColumnTransformer(transformers=[('num', numericas_pipe, numericas_pipe_nombres),
                                                      ('ord', ordinales_pipe, ordinales_pipe_nombres),
                                                      ('nom', nominales_pipe, nominales_pipe_nombres)], remainder='passthrough')

# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 6**

In [24]:
# Como se va a utilizar Validación-Cruzada, concatena los conjuntos de entrenamiento y prueba
# en uno nuevo conjunto aumentado que llamaremos trainval para utilizar como entrenamiento:

# ************* Inlcuye aquí tu código:**************************

Xtraintest = pd.concat([Xtrain, Xtest], axis=0)
ytraintest = pd.concat([ytrain, ytest], axis=0)

# *********** Aquí termina la sección de agregar código *************

# Veamos cuántas variables nuevas se introducen con las transformaciones One-Hot-Encoding:
Xtmp = Xtraintest.copy()
tmp = columnasTransformer.fit_transform(Xtmp)
print("Dimensión de las variables de entrada ANTES de las transformaciones:", Xtmp.shape)
print("Dimensión de las variables de entrada DESPUÉS de las transformaciones:", tmp.shape)

Dimensión de las variables de entrada ANTES de las transformaciones: (1000, 20)
Dimensión de las variables de entrada DESPUÉS de las transformaciones: (1000, 41)


## **Ejercicio 7**

In [25]:
# Definimos a continuación la función que llamamos "mis_modelos" que incluye
# todos los modelos que deseamos comparar en el ejercicio.


def mis_modelos():
  modelos, nombres = list(), list()


  # ************* Inlcuye aquí tu código:**************************
  #
  # Deberás incluir en cada modelo los argumentos que consideres
  # adecuados para que cada uno converja y no esté sobreentrenado
  # con respecto a la métrica de la exatitud (accuracy).


  # Regresión Logística - Logistic Regression-LR:
  # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

  modelos.append(LogisticRegression(penalty='elasticnet',
                                    solver = 'saga',
                                    tol=0.0001,     # tolerancia
                                    fit_intercept = True,
                                    max_iter = 2000,
                                    C = 10,
                                    l1_ratio= 0.34,
                                    random_state=1))
  nombres.append('LR')




  # k-Vecinos más Cercanos : k-Nearest-Neighbors-kNN:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html

  modelos.append(KNeighborsClassifier(n_neighbors=20,
                                      p = 1,
                                      leaf_size = 35,
                                      n_jobs = -1,
                                      weights = "uniform") )
  nombres.append('kNN')




  # Árbol de decisiones-DecisionTree-DT:
  # https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html

  modelos.append(DecisionTreeClassifier(criterion = "entropy",
                                        max_depth = 10,
                                        min_samples_split = 15,
                                        random_state = 1 ))
  nombres.append('DTree')




  # Bosque Aleatorio-RandomForest-RF:
  # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

  modelos.append(RandomForestClassifier(n_estimators = 200,
                                        criterion = "entropy",
                                        max_features = "sqrt",
                                        random_state =  1,
                                        max_depth=10))
  nombres.append('RF')




  # XGBoosting:
  # https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.XGBClassifier
  # https://xgboost.readthedocs.io/en/stable/parameter.html

  modelos.append(XGBClassifier(random_state  = 1 ,
                               tree_method="hist",
                               learning_rate=0.05,
                               max_depth=7))
  nombres.append('XGBoost')




  # Red neuronal de Perceptrón Multicapa-MLP:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html

  modelos.append(MLPClassifier(solver = "adam",
                               random_state  = 1 ,
                               hidden_layer_sizes = 32,
                               activation = "identity",
                               max_iter = 2500,
                               alpha = 0.001))
  nombres.append('MLP')




  # Máquina de Vectores de Soporte-SVM:
  # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

  modelos.append(SVC(C = 14,
                     kernel = "linear",
                     tol = 0.0001))
  nombres.append('SVM')

  return modelos, nombres


# Técnica de submuestreo (undersampling) y/o sobremuestreo (versampling) utilizada:
mi_uoSampling = KMeansSMOTE(random_state=1, k_neighbors = 6, n_jobs = -1) ##BorderlineSMOTE(random_state=1, k_neighbors = 10) #SMOTE(random_state=1, k_neighbors = 2)

Xtv_uo, ytv_uo = mi_uoSampling.fit_resample(Xtraintest, ytraintest)



# *********** Aquí termina la sección de agregar código *******************





# Entrenamos cada uno de los modelos y desplegamos la métricas de Train y Val.

# NOTA: Observa que el método de Validación-Cruzada llama  a los resultados
#       de "validation" como "test":

modelos, nombres = mis_modelos()
resultados = list()

for i in range(len(modelos)):

  # Definimos nuestro pipeline con las transformaciones y los modelos:
  pipeline = Pipeline(steps=[('ct',columnasTransformer),('m',modelos[i])])

  # Aplicaremos validación-cruzada:
  micv = RepeatedStratifiedKFold(n_splits=5,
                                 n_repeats=3,
                                 random_state=5     # agreguemos una semilla para estabilizar resultados.
                                 )


  # Definimos las métricas que desamos recuperar:
  mismetricas = {'miaccuracy':'accuracy','miprecision':'precision','mirecall':'recall',
                 'mifi':'f1','miauc':'roc_auc','migmean':make_scorer(geometric_mean_score)}

  # Llevamos a cabo el entrenamiento:
  scores = cross_validate(pipeline,
                          Xtv_uo,
                          ytv_uo,
                          scoring=mismetricas,
                          cv=micv,
                          return_train_score=True,
                          )

  # Guardemos el resultado de cada modelo para análisis posteriores.
  resultados.append(scores)

  # Desplegamos los valores de las métricas para verificar si no hay
  # subentrenamiento o sobreentrenamiento:
  print('>> %s' % nombres[i])
  for j,k in enumerate(list(scores.keys())):
    if j>1:
      print('\t %s %.3f (%.3f)' % (k, np.mean(scores[k]),np.std(scores[k])))

>> LR
	 test_miaccuracy 0.817 (0.029)
	 train_miaccuracy 0.832 (0.007)
	 test_miprecision 0.827 (0.034)
	 train_miprecision 0.845 (0.010)
	 test_mirecall 0.803 (0.036)
	 train_mirecall 0.813 (0.007)
	 test_mifi 0.814 (0.030)
	 train_mifi 0.829 (0.007)
	 test_miauc 0.903 (0.021)
	 train_miauc 0.921 (0.005)
	 test_migmean 0.816 (0.030)
	 train_migmean 0.832 (0.007)
>> kNN
	 test_miaccuracy 0.795 (0.020)
	 train_miaccuracy 0.817 (0.006)
	 test_miprecision 0.840 (0.022)
	 train_miprecision 0.865 (0.011)
	 test_mirecall 0.731 (0.032)
	 train_mirecall 0.751 (0.008)
	 test_mifi 0.781 (0.023)
	 train_mifi 0.804 (0.006)
	 test_miauc 0.881 (0.022)
	 train_miauc 0.907 (0.005)
	 test_migmean 0.793 (0.021)
	 train_migmean 0.814 (0.006)
>> DTree
	 test_miaccuracy 0.792 (0.026)
	 train_miaccuracy 0.888 (0.009)
	 test_miprecision 0.802 (0.034)
	 train_miprecision 0.910 (0.021)
	 test_mirecall 0.777 (0.039)
	 train_mirecall 0.862 (0.030)
	 test_mifi 0.789 (0.027)
	 train_mifi 0.885 (0.010)
	 test_miauc

## **Ejercicio 8**

### **Escribe tus conclusiones finales de la actividad.**

### ++++++++ Inicia la sección de agregar texto: +++++++++++


### **Conclusiones Finales de la Actividad** :

En esta actividad se analizaron diferentes modelos y el uso de submuestreo o sobre-muestreo para casos donde las clases no estan balanceadas.

Entre las ventajas del sobre-muestreo esta que no se pierde información, en el caso de dataset pequeños previene que se pierdan datos de la clase mayoritaria. Este tipo de técnica agrega datos sintéticos a la clase minoritaria. Sin embargo, al aumentar el conjunto de datos los tiempos de entrenamiento y uso de memoria serán mayores, además de un riesgo a sobreajuste del modelo.

En el caso del submuestreo, se reduce el tamaño del dataset, lo que beneficia los tiempos de entrenaamiento y reduce a su vez el uso dde memoria. Para grandes volúmenes de datos se puede utilizar ya que no hay pérdida de datos significativa. Aun así esta el riesgo de perder información valiosa, por lo que no se sugiere para conjuntos de datos pequeños.


Los modelos implementados fueron:


* **Logistic regression**: Este algoritmo de aprendizaje supervisado es utilizado para problemas de clasificcación, predice las probabilidades de un objetivo binario según las variables predictoras (X) realizando un ajuste lineal con una salida de 0 o 1.

* **K-Nearest Neighbors (KNN)**:Este algoritmo es utilizado para análisis de clasificación y regresión lineal. Predice la variable de salida basada en las K vecinas más cercanas en el conjunto de entrenamiento. El valor de k determina el número de vecinos a considerar a la hora de hacer la predicción.

* **Modelo de árbol de decisión**: También se trata de un algoritmo para análisis de clasificacción y regresión. Divide de forma recurrente los datos en subconjuntos más pequeños para crear el arbol de decisión.

* **Random Forest**: Este algoritmo ajusta una serie de clasificadores de árboles de decisión en varias submuestras y utiliza promedios para mejorar la precisión y controlar el sobreajuste.

* **Extreme Gradient Boosting (XGBoost)**: O refuerzo de gradientes extremo, es un método basado en árboles de decisión. Realiza una predicción inicial y crea un arbol de decisión, la salida del árbol se utiliza para construir otro árbol  hasta que los residuales dejan de reducirse, o bien el número de veces especificado.

* **Red neuronal de Perceptrón Multicapa-MLP**: Las redes neuronales son modelos inspirados en el cerebro humano, consisten en multiples capas con varias neuronas cada una, capaces de aprender funciones complejas.

* **Support Vector Machines (SVM)**: Este algoritmo es usado principalmente para problemas de clasificación pero tambien puede usarse en problemas de regresión. Clasifica los datos encontrando el hiperplano óptimo que separa al máximo las diferentes clases en los datos de entrada.


### **Análisis de Resultados**
A continuación, se presentan las métricas clave obtenidas para cada modelo en los conjuntos de prueba y entrenamiento:

1. **Regresión Logística (LR)**:
   - G-mean Test: 0.816 (0.030)
   - G-mean Train: 0.832 (0.007)

   **Interpretación**, El modelo muestra un buen desempeño. la diferencia es minima lo que indica una buena generalización y ausencia de sobreajuste significativo.

2. **k-Vecinos más Cercanos (kNN)**:
   - G-mean Test: 0.793 (0.021)
   - G-mean Train: 0.814 (0.006)

   **Interpretación**,Existe una diferencia notable entre las métricas de entrenamiento y prueba, indicando un sobreajuste y menor capacidad de generalización.

3. **Árbol de Decisión (DTree)**:
   - G-mean Test: 0.791 (0.026)
   - G-mean Train: 0.887 (0.009)

   **Intepretación**, En este modelo se puede observar una notable diferencia entre Train y Test, indicando un sobreajuste. Nos indica que el modelo de árbol de desición esta aprendiendo demasiado del conjunto de Train.

4. **Bosque Aleatorio (RF)**:
   - G-mean Test: 0.813 (0.028)
   - G-mean Train: 0.970 (0.005)

   **Interpretación**,Las métricas son ligeramente superiores a las del Árbol de Decisión, indicando que el ensamble mejora el desempeño, pero aún hay espacio para optimización. Pero de igual manera nos muestra un sobreajuste con un valor de G-mean Train demasiado alto comparado con la Test.

5. **XGBoost**:
   - G-mean Test: 0.816 (0.023)
   - G-mean Train: 0.970 (0.005)
   
   **Interpretación**, Aunque presenta el mejor desempeño en prueba, existe una diferencia con el entrenamiento sugiere un gran sobreajuste significativo.

6. **Perceptrón Multicapa (MLP)**:
   - G-mean Test: 0.813 (0.026)
   - G-mean Train: 0.831 (0.008)
   
   **Interpretación**,Este modelo nos muestra un leve sobreajuste, mostrando un buen equilibrio entre el conjunto de Entrenamiento y Prueba.

7. **Máquina de Vectores de Soporte (SVM)**:
   - G-mean Test: 0.822 (0.030)
   - G-mean Train: 0.840 (0.007)

   **Interpretación**, este modelo representa un buen desempeño con el conjunto Prueba, con una diferencia muy minima con el Entrenamiento.

###**Mejor Modelo**
En conclusión, los modelos con **Mejor Rendimiento** y demostrando ser los modelos más robustos son los siguientes:

 -**Maquina de Vectores de Soporte (SVM)**

 -**Regresión Logistica (LR-Elastic)**


### **Conclusión General**

La actividad permitió comparar múltiples modelos y técnicas para abordar un problema de clasificación con datos desbalanceados en el contexto del riesgo crediticio. Aunque XGBoost mostró el mejor desempeño en términos de métricas en el conjunto de prueba, su tendencia al sobreajuste sugiere que no es la opción más confiable sin ajustes adicionales.

La Máquina de Vectores de Soporte (SVM) y la Regresión Logística emergen como los modelos más balanceados, ofreciendo un buen desempeño y capacidad de generalización. Esto resalta la importancia de considerar no solo las métricas de desempeño sino también la estabilidad y robustez del modelo al momento de seleccionar la mejor opción para implementación.

Finalmente, la experiencia destaca la relevancia de manejar adecuadamente el desbalance de clases y de utilizar técnicas de validación apropiadas para obtener modelos confiables y aplicables en contextos reales.



**Referencias**

Huyen, C. (s. f.). Designing machine learning systems. O’Reilly Online Learning. https://learning.oreilly.com/library/view/designing-machine-learning/9781098107956/ch04.html

Abhishek, K., & Abdelaziz, M. (s. f.). Machine Learning for Imbalanced Data. O’Reilly Online Learning. https://learning.oreilly.com/library/view/machine-learning-for/9781801070836/B17259_01.xhtml#_idParaDest-20

RandomForestClassifier. (s. f.). Scikit-learn. https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.RandomForestClassifier.html

Cómo funciona el algoritmo XGBoost—ArcGIS Pro | Documentación. (s. f.). https://pro.arcgis.com/es/pro-app/latest/tool-reference/geoai/how-xgboost-works.htm#:~:text=XGBoost%20es%20la%20abreviatura%20de,aleatorio%20y%20refuerzo%20de%20gradientes.

XGBoost Documentation — xgboost 2.1.1 documentation. (s. f.). https://xgboost.readthedocs.io/en/stable/
### ++++++++ Termina la sección de agregar texto. +++++++++++


>> ### **Fin de la Actividad de las Semanas 5 y 6.**